pkg/metrics/sampler/matcher.go

```diff
@@ -0,0 +1,209 @@
+	// regex should never be nil here.
+	regex := regexCache[expected.(string)]
+	return regex.MatchString(fmt.Sprintf("%v", actual))
+}
+
+//newExpressionMatcher returns a new ExpressionMatcher
+func newExpressionMatcher(dimensionName string, expr string) ExpressionMatcher {
+	return build(dimensionName, expr)
+}
+
+func build(dimensionName string, expr string) ExpressionMatcher {
+	// if the dimension is not "registered", return a constant "false" matcher
+	// "false" will make the chain continue (until either there is a "true" result or there's no more matchers),
+	// so this matcher basically get's ignored in the current implementation
+	mappedAttributeName, found := attrCache[dimensionName]
+	if !found {
+		return constantMatcher{false}
+	}
+
+	eval := matcher{
+		PropertyName: mappedAttributeName,
+	}
+
+	if strings.HasPrefix(expr, "regex") {
+		regex := strings.Trim(strings.TrimSpace(strings.TrimLeft(expr, "regex")), `"`)
+		cacheRegex(regex)
+		eval.ExpectedValue = regex
+		eval.Evaluator = regularExpressionEvaluator
+	} else {
+		eval.ExpectedValue = strings.TrimSpace(strings.Trim(expr, `"`))
+		eval.Evaluator = literalExpressionEvaluator
+	}
+
+	return eval
+}
+
+func cacheRegex(regex string) {
+	//if not cached yet, cache it
+	if _, ok := regexCache[regex]; !ok {
+		regexCache[regex] = regexp.MustCompile(regex)
+	}
+}
+
+// MatcherChain is a chain of evaluators
+// An evaluator chain stores for each attribute an array of evaluators
+// Each evaluator represent a single rule that evaluates against the attribute
+// Usually each attribute will have a single evaluator.
+// For example:
+// - process.name
+//   - "test"
+// - process.executable
+//   - "/bin/test"
+//   - regex "^/opt/newrelic/"
+// will create an evaluator chain with 2 entries. The first one will have 1 evaluator. The second 2 evaluators
+type MatcherChain struct {
+	Matchers map[string][]ExpressionMatcher
+	Enabled  bool
+}
+
+// NewMatcherChain creates a new chain of matchers.
+// Each expression will generate an matcher that gets added to the chain
+// While the chain will be matched for each "sample", it terminates as soon as 1 match is matched (result = true)
+func NewMatcherChain(expressions map[string][]string) MatcherChain {
+	chain := MatcherChain{Matchers: map[string][]ExpressionMatcher{}, Enabled: false}
+
+	// no matchers means the chain will be disabled
+	if len(expressions) == 0 {
+		return chain
+	}
+
+	chain.Enabled = true
+	for prop, exprs := range expressions {
+		if _, ok := chain.Matchers[prop]; !ok {
+			chain.Matchers[prop] = []ExpressionMatcher{}
+		}
+
+		evs := chain.Matchers[prop]
+		for _, expr := range exprs {
+			e := newExpressionMatcher(prop, expr)
+			evs = append(evs, e)
+		}
+		chain.Matchers[prop] = evs
+	}
+
+	return chain
+}
+
+// Evaluate returns the result of evaluating a chain of evaluators
+func (ec MatcherChain) Evaluate(event interface{}) bool {
+	// default is to return true. what "true" means is up to the client
```

Comment:
I guess we could define a cleaner contract by returning a bool pointer (or second bool return) where nil (or second return) says there's no match